Upgrade to Uno 3.x, minor control changes required

mitchcapper opened the pull request:
<!-- Please target use the correct target branch for your PR (3.x for LibVLCSharp 3, current master is for LibVLCSharp/LibVLC 4). -->

Switched to Uno 3.x.  Using an Uno 2.x control in an Uno 3.x project can cause odd crash exceptions.  Officially they say 3.x is soruce compat with 2.x so people shouldn't have an issue upgrading but clearly there can be minor differences.  

<!-- Describe your changes here. -->

### Issues Resolved ### 
fixes #375

### API Changes ###
None

### Platforms Affected ### 
Core

### Behavioral/Visual Changes ###
This moves users to uno 3.x so while there are no libvlcsharp changes there are some minor upgrade notes:
https://github.com/unoplatform/uno/blob/master/doc/articles/migrating-from-previous-releases.md
One tooltip had to migrate from content presenter to content control however I don't believe it should cause an issue.  This is due to CP now allowing to have a content property bound to a non template item.

### Before/After Screenshots ### 
No UI changes.

### Testing Procedure ###
Tested Android and UWP no ios device.

### PR Checklist ###

- [X ] Rebased on top of the target branch at time of PR
- [ X] Changes adhere to coding standard

relevant links:
- https://github.com/unoplatform/uno/blob/master/doc/articles/migrating-from-previous-releases.md
- https://github.com/unoplatform/uno/releases/tag/3.0.5

mfkl commented:
Hi,

Thanks for the PR!

It seems the CI is complaining about this:
```
2020-09-23T16:33:52.2081909Z "D:\a\1\s\src\LibVLCSharp.sln" (Build target) (1) ->
2020-09-23T16:33:52.2082790Z "D:\a\1\s\src\LibVLCSharp.Uno\LibVLCSharp.Uno.csproj" (default target) (30) ->
2020-09-23T16:33:52.2084246Z "D:\a\1\s\src\LibVLCSharp.Uno\LibVLCSharp.Uno.csproj" (Build target) (30:6) ->
2020-09-23T16:33:52.2086410Z   obj\Release\monoandroid81\g\XamlCodeGenerator\VideoView_Android_iOS_0654003b6e8285e8171ea5eceba6aa4a.g.cs(118,4): error CS0029: Cannot implicitly convert type 'Windows.UI.Xaml.Controls.Border' to 'Android.Views.View' [D:\a\1\s\src\LibVLCSharp.Uno\LibVLCSharp.Uno.csproj]
2020-09-23T16:33:52.2094712Z   obj\Release\monoandroid81\g\XamlCodeGenerator\MediaPlayerElement_44b984361be691e5a697b2e5e83c7ee0.g.cs(118,4): error CS0029: Cannot implicitly convert type 'Windows.UI.Xaml.Controls.Grid' to 'Android.Views.View' [D:\a\1\s\src\LibVLCSharp.Uno\LibVLCSharp.Uno.csproj]
2020-09-23T16:33:52.2097583Z   VideoView.Android.cs(17,71): error CS0103: The name 'ContextHelper' does not exist in the current context [D:\a\1\s\src\LibVLCSharp.Uno\LibVLCSharp.Uno.csproj]
2020-09-23T16:33:52.2101171Z   VideoView.Android.cs(18,21): error CS1061: 'Border' does not contain a definition for 'AddChild' and no accessible extension method 'AddChild' accepting a first argument of type 'Border' could be found (are you missing a using directive or an assembly reference?) [D:\a\1\s\src\LibVLCSharp.Uno\LibVLCSharp.Uno.csproj]
2020-09-23T16:33:52.2103330Z   obj\Release\monoandroid81\g\XamlCodeGenerator\VideoView_Android_iOS_0654003b6e8285e8171ea5eceba6aa4a.g.cs(64,76): error CS0029: Cannot implicitly convert type 'Android.Views.View' to 'Windows.UI.Xaml.UIElement' [D:\a\1\s\src\LibVLCSharp.Uno\LibVLCSharp.Uno.csproj]
2020-09-23T16:33:52.2105482Z   obj\Release\monoandroid81\g\XamlCodeGenerator\VideoView_Android_iOS_0654003b6e8285e8171ea5eceba6aa4a.g.cs(64,76): error CS1662: Cannot convert lambda expression to intended delegate type because some of the return types in the block are not implicitly convertible to the delegate return type [D:\a\1\s\src\LibVLCSharp.Uno\LibVLCSharp.Uno.csproj]
2020-09-23T16:33:52.2108900Z   obj\Release\monoandroid81\g\XamlCodeGenerator\PlaybackControls_a640c024b4ec96fcd73a42d435df6be4.g.cs(69,76): error CS0029: Cannot implicitly convert type 'Android.Views.View' to 'Windows.UI.Xaml.UIElement' [D:\a\1\s\src\LibVLCSharp.Uno\LibVLCSharp.Uno.csproj]
2020-09-23T16:33:52.2111046Z   obj\Release\monoandroid81\g\XamlCodeGenerator\PlaybackControls_a640c024b4ec96fcd73a42d435df6be4.g.cs(69,76): error CS1662: Cannot convert lambda expression to intended delegate type because some of the return types in the block are not implicitly convertible to the delegate return type [D:\a\1\s\src\LibVLCSharp.Uno\LibVLCSharp.Uno.csproj]
2020-09-23T16:33:52.2113242Z   obj\Release\monoandroid81\g\XamlCodeGenerator\MediaPlayerElement_44b984361be691e5a697b2e5e83c7ee0.g.cs(64,76): error CS0029: Cannot implicitly convert type 'Android.Views.View' to 'Windows.UI.Xaml.UIElement' [D:\a\1\s\src\LibVLCSharp.Uno\LibVLCSharp.Uno.csproj]
2020-09-23T16:33:52.2115519Z   obj\Release\monoandroid81\g\XamlCodeGenerator\MediaPlayerElement_44b984361be691e5a697b2e5e83c7ee0.g.cs(64,76): error CS1662: Cannot convert lambda expression to intended delegate type because some of the return types in the block are not implicitly convertible to the delegate return type [D:\a\1\s\src\LibVLCSharp.Uno\LibVLCSharp.Uno.csproj]
2020-09-23T16:33:52.2118689Z   obj\Release\monoandroid81\g\XamlCodeGenerator\PlaybackControls_a640c024b4ec96fcd73a42d435df6be4.g.cs(219,74): error CS0029: Cannot implicitly convert type 'Android.Views.View' to 'Windows.UI.Xaml.UIElement' [D:\a\1\s\src\LibVLCSharp.Uno\LibVLCSharp.Uno.csproj]
2020-09-23T16:33:52.2120944Z   obj\Release\monoandroid81\g\XamlCodeGenerator\PlaybackControls_a640c024b4ec96fcd73a42d435df6be4.g.cs(219,74): error CS1662: Cannot convert lambda expression to intended delegate type because some of the return types in the block are not implicitly convertible to the delegate return type [D:\a\1\s\src\LibVLCSharp.Uno\LibVLCSharp.Uno.csproj]
2020-09-23T16:33:52.2124460Z   obj\Release\monoandroid81\g\XamlCodeGenerator\PlaybackControls_a640c024b4ec96fcd73a42d435df6be4.g.cs(241,74): error CS0029: Cannot implicitly convert type 'Android.Views.View' to 'Windows.UI.Xaml.UIElement' [D:\a\1\s\src\LibVLCSharp.Uno\LibVLCSharp.Uno.csproj]
2020-09-23T16:33:52.2127310Z   obj\Release\monoandroid81\g\XamlCodeGenerator\PlaybackControls_a640c024b4ec96fcd73a42d435df6be4.g.cs(241,74): error CS1662: Cannot convert lambda expression to intended delegate type because some of the return types in the block are not implicitly convertible to the delegate return type [D:\a\1\s\src\LibVLCSharp.Uno\LibVLCSharp.Uno.csproj]
2020-09-23T16:33:52.2130747Z   obj\Release\monoandroid81\g\XamlCodeGenerator\PlaybackControls_a640c024b4ec96fcd73a42d435df6be4.g.cs(5965,5): error CS0029: Cannot implicitly convert type 'Windows.UI.Xaml.Controls.Border' to 'Android.Views.View' [D:\a\1\s\src\LibVLCSharp.Uno\LibVLCSharp.Uno.csproj]
2020-09-23T16:33:52.2132505Z   obj\Release\monoandroid81\g\XamlCodeGenerator\PlaybackControls_a640c024b4ec96fcd73a42d435df6be4.g.cs(4036,76): error CS0029: Cannot implicitly convert type 'Android.Views.View' to 'Windows.UI.Xaml.UIElement' [D:\a\1\s\src\LibVLCSharp.Uno\LibVLCSharp.Uno.csproj]
2020-09-23T16:33:52.2134653Z   obj\Release\monoandroid81\g\XamlCodeGenerator\PlaybackControls_a640c024b4ec96fcd73a42d435df6be4.g.cs(4036,76): error CS1662: Cannot convert lambda expression to intended delegate type because some of the return types in the block are not implicitly convertible to the delegate return type [D:\a\1\s\src\LibVLCSharp.Uno\LibVLCSharp.Uno.csproj]
2020-09-23T16:33:52.2146248Z   obj\Release\monoandroid81\g\XamlCodeGenerator\PlaybackControls_a640c024b4ec96fcd73a42d435df6be4.g.cs(5923,6): error CS0029: Cannot implicitly convert type 'Windows.UI.Xaml.Shapes.Ellipse' to 'Android.Views.View' [D:\a\1\s\src\LibVLCSharp.Uno\LibVLCSharp.Uno.csproj]
2020-09-23T16:33:52.2148719Z   DisplayRequestAdapter.cs(34,13): error Uno0001: Windows.System.Display.DisplayRequest.RequestRelease() is not implemented in Uno [D:\a\1\s\src\LibVLCSharp.Uno\LibVLCSharp.Uno.csproj]
2020-09-23T16:33:52.2150318Z   DisplayRequestAdapter.cs(26,13): error Uno0001: Windows.System.Display.DisplayRequest.RequestActive() is not implemented in Uno [D:\a\1\s\src\LibVLCSharp.Uno\LibVLCSharp.Uno.csproj]
2020-09-23T16:33:52.2152163Z   DisplayRequestAdapter.cs(16,30): error Uno0001: Windows.System.Display.DisplayRequest is not implemented in Uno [D:\a\1\s\src\LibVLCSharp.Uno\LibVLCSharp.Uno.csproj]
```
https://videolan.visualstudio.com/73ec5719-cee8-4da9-937a-9d8a3993d7f5/_apis/build/builds/3389/logs/26

>  Using an Uno 2.x control in an Uno 3.x project can cause odd crash exceptions.

Does this mean we can only support 1 version of Uno within one package?

mitchcapper commented:
I had a heck of a time getting it to build (just the library in general).  I even got some of these similar errors.   Eventually to get it to build I was using android 9 and unloaded virtually all the projects other than UWP and android.   After going through a slew of changes I tried reverting as much as I could to get it to still work.

I did find now I missed the IOS add so that should fix a few of the errors.

I also got the DisplayRequest errors, which I fixed by making it public rather than internal.   As that was all libvlc code I couldn't figure out why that would be a new error.   The assembly was properly marked as friend as well.   I reverted it back to internal at the end and rebuilt and it was fine.

I am not sure if it just happens when there is not a successful build or if I am missing something.

Looking at the other errors it is likely relating to some of the templating child fetching that macos seems to catch but android doesn't oddly.   

I will go through the theming and see.  The View is no longer directly assigned as a child so may need to alter how they are grabbing the VideoView.

Otherwise I can see about installing the android 8 sdk to see if that is an issue, or it could also be IOS related (as the windows does the cross building).

In theory the VideoView.Android should only be compiled in for android and ios for ios right?   As the cross compile errors seem odd.   I could further gate them with platform #ifdefs but I didn't do so as I figured each was only used for its specific case.

mfkl commented:
> I had a heck of a time getting it to build (just the library in general). 

Sorry to hear that. Have you followed our contribution guide and had a look at what the CI does and how it is setup? This can be helpful to have the solution building quicker.

> In theory the VideoView.Android should only be compiled in for android and ios for ios right? 

Yes.
```
  <ItemGroup Condition="$(TargetFramework.StartsWith('MonoAndroid'))">
    <ProjectReference Include="..\LibVLCSharp.Android.AWindow\LibVLCSharp.Android.AWindow.csproj" PrivateAssets="All" />
    <Compile Include="**\*.Android.*cs" />
    <Page Include="**\*.Android.*xaml" Exclude="bin\**\*.xaml;obj\**\*.xaml" />
  </ItemGroup>
  <ItemGroup Condition="$(TargetFramework.StartsWith('Xamarin.iOS'))">
    <Compile Include="**\*.iOS.*cs" />
    <Page Include="**\*.iOS.*xaml" Exclude="bin\**\*.xaml;obj\**\*.xaml" />
  </ItemGroup>
```

mfkl commented on src/LibVLCSharp.Uno/Themes/PlaybackControls.xaml:
```diff
@@ -262,7 +262,7 @@
                                                             <ToolTipService.ToolTip>
                                                                 <ToolTip x:Name="ThumbnailTooltip">
-                                                                    <ContentPresenter Content="{Binding}" />
+                                                                    <ContentControl Content="{Binding}" />
```
Are you sure about this change?

mitchcapper replied:
ok I give up I can't figure out where this tooltip should show up.  I have built straight 3.x on uno 2 as well and can't quite find what slider this references.    Is this just the position ellipse?  If so what should the tooltip be displaying?
![image](https://user-images.githubusercontent.com/1643324/122036801-d18f2e00-cd88-11eb-873c-0a6503f92d15.png)

mfkl commented:
> Does this mean we can only support 1 version of Uno within one package?

![tweet-uno](https://user-images.githubusercontent.com/3928834/94156623-935f8080-fe80-11ea-811a-14e825763bee.PNG)

That settles it. We will moving to Uno 3, I guess.

mitchcapper commented:
The other option would be to create an uno 2.x legacy package or new uno 3 package, but in theory as the upgrade to uno 3 shouldn't be too big for users that may just be the logistically simpler option.

I will test the playback controls and further theme issues this weekend hopefully.  From what I read that was the correct direction to be going for the sort of binding done.

As for build issues I am a big fan of docker and CI setup scripts and the failures I am sure are my own.  A bit busy with work so only had so much time to see if I could fix this (just using libvlc for a side project).

I will try and look at it in more detail to see if there is anything I could think to add that might help other users (some of it is just opaque error messaging from the build system so not much to do about that).   I think the solution itself is logistically complex with all the platforms so likely i will RTFM and just see if I can put some more warnings or help for what some errors may actually mean.

mfkl commented:
> The other option would be to create an uno 2.x legacy package or new uno 3 package

Yes, but I don't think it is worth the overhead with packaging, given how few users we seem to have on Uno.

> As for build issues I am a big fan of docker

So am I, but I use it for libvlc (https://code.videolan.org/videolan/docker-images), not _libvlcsharp_ (though it has been planned https://code.videolan.org/videolan/LibVLCSharp/-/issues/91).

> I will try and look at it in more detail to see if there is anything I could think to add that might help other users (some of it is just opaque error messaging from the build system so not much to do about that). I think the solution itself is logistically complex with all the platforms so likely i will RTFM and just see if I can put some more warnings or help for what some errors may actually mean.

You need all the proper SDKs (Xamarin, netcore, desktop, etc.). Aside from this, the great [MSBuildSdkExtras](MSBuildSdkExtras) that we use for multi targeting should do the job (it is pulled automagically from the csproj).

mitchcapper commented:
Don't merge this in yet (if you ever would) still want to test some of the templating things.  The error was due to not updating to Android 9 as Uno requires (this may also stop you from merging).

mfkl commented on src/LibVLCSharp.Uno/LibVLCSharp.Uno.csproj:
```diff
@@ -5,7 +5,7 @@
     <LangVersion>8.0</LangVersion>
     <Nullable>enable</Nullable>
-    <TargetFrameworks>Xamarin.iOS10;MonoAndroid81</TargetFrameworks>
+    <TargetFrameworks>Xamarin.iOS10;MonoAndroid90</TargetFrameworks>
```
Would it still work if we only target Android 9 in LibVLCSharp.Uno and remain on Android 8.1 in the other projects?

mfkl replied:
Seems to work.

mfkl commented:
> Don't merge this in yet (if you ever would) still want to test some of the templating things.

Ok. Yes, I intend to merge your contribution :-)

You can only test UWP and Android, right? Let me know and I'll check iOS.
We also need to make sure the media element is still OK.

mfkl commented:
Feel free to let me know when this is ready @mitchcapper

MartinZikmund commented:
@mfkl @mitchcapper Can this be reopened and finalized please? It would be very beneficial to bring support for latest versions of Uno Platform.

jeremyVignelles commented:
@MartinZikmund : Feel free to open your own PR to finish the job, I guess :)

ahoefling commented:
@MartinZikmund I would be more than happy to help you look at this. We are going to need support for libvlc with Uno and Android later this year (Q3/Q4 timeline). I am more than happy to donate my time as well as some developers on my team to get this working.

jeremyVignelles commented:
Good to hear, feel free to join the Discord server if you need anything and to synchronize your work 😉

mitchcapper commented on src/LibVLCSharp.Uno/Themes/PlaybackControls.xaml:
```diff
@@ -353,7 +353,6 @@
                                                     VerticalAlignment="Stretch">
                                                 <ContentPresenter ContentTemplate="{TemplateBinding ContentTemplate}"
                                                                   ContentTransitions="{TemplateBinding ContentTransitions}"
-                                                                  Content="{TemplateBinding Content}"
```
I believe this is correct that ContentPresenter should bind to the Content property by default so removing is not an issue.  Technically I think we can remove ContentTemplate  as well as this should be default.   If we used a different parameter base setting ContentSource allows renaming the binding on those two by default.

mitchcapper replied:
As I understand this styles the volume flyout it seems to work with the change: 
![image](https://user-images.githubusercontent.com/1643324/122033369-4f513a80-cd85-11eb-8f3b-bfc8873b4b52.png)

mfkl replied:
👍

mitchcapper commented:
NOTE! This is not yet ready for merge the templating I am still testing and playing with.  From my original PR here are the changes:
*) Targetting the latest uno 3.7.3
*) Android version bumped up to 10 on all android projects.  Android 10 is the minimum supported version for google now.  Uno did the same (so 10 is minimum for current uno version).   If we didn't want to do so the non-uno projects could target a lower android framework likely (let me know and I can revert to 81).
*) The original debate item, ContentPresenter.   What VS will now warn or error over is that binding to the content property of it like we were is likely not the right way to be using it (and can result in undefined behavior they warn on).   I will still test and validate the change direction here is correct but either the ContentSource should be used or contentcontrol for actual binding from what I have found thus far.  If It tests properly though I would assume changes are fine.

mfkl commented:
> Android version bumped up to 10 on all android projects. Android 10 is the minimum supported version for google now. Uno did the same (so 10 is minimum for current uno version). If we didn't want to do so the non-uno projects could target a lower android framework likely (let me know and I can revert to 81).

Yes please, I'd like to stick to 8.1 for now if possible.

mfkl commented:
If anyone is still interested to take this up, much appreciated. Will allow us to remove a new needed workaround https://github.com/videolan/libvlcsharp/commit/00952b64c57e1e2cc65e9a2e4b417e797654a1ff

mitchcapper commented:
So I can't test this as I upgraded VS and ran into a bug, it seems the autobuild system may be hitting that same bug with the latest vs tools.
```
 MSBUILD : error : Generation failed: System.MissingMethodException: Method not found: 'System.Collections.Immutable.ImmutableList`1<!!0> System.Collections.Immutable.ImmutableList.ToImmutableList(Builder<!!0>)'. [D:\a\1\s\src\LibVLCSharp.Uno\LibVLCSharp.Uno.csproj]
MSBUILD : error :    at Microsoft.Build.Evaluation.LazyItemEvaluator`4.UpdateOperation.ApplyImpl(Builder listBuilder, ImmutableHashSet`1 globsToIgnore) [D:\a\1\s\src\LibVLCSharp.Uno\LibVLCSharp.Uno.csproj]
MSBUILD : error :    at Microsoft.Build.Evaluation.LazyItemEvaluator`4.LazyItemOperation.Apply(Builder listBuilder, ImmutableHashSet`1 globsToIgnore) [D:\a\1\s\src\LibVLCSharp.Uno\LibVLCSharp.Uno.csproj]
MSBUILD : error :    at Microsoft.Build.Evaluation.LazyItemEvaluator`4.LazyItemList.ComputeItems(LazyItemList lazyItemList, ImmutableHashSet`1 globsToIgnore) [D:\a\1\s\src\LibVLCSharp.Uno\LibVLCSharp.Uno.csproj]
MSBUILD : error :    at Microsoft.Build.Evaluation.LazyItemEvaluator`4.LazyItemList.GetItemData(ImmutableHashSet`1 globsToIgnore) [D:\a\1\s\src\LibVLCSharp.Uno\LibVLCSharp.Uno.csproj]
MSBUILD : error :    at System.Linq.Enumerable.<SelectManyIterator>d__17`2.MoveNext() [D:\a\1\s\src\LibVLCSharp.Uno\LibVLCSharp.Uno.csproj]
MSBUILD : error :    at System.Linq.Buffer`1..ctor(IEnumerable`1 source) [D:\a\1\s\src\LibVLCSharp.Uno\LibVLCSharp.Uno.csproj]
MSBUILD : error :    at System.Linq.OrderedEnumerable`1.<GetEnumerator>d__1.MoveNext() [D:\a\1\s\src\LibVLCSharp.Uno\LibVLCSharp.Uno.csproj]
MSBUILD : error :    at Microsoft.Build.Evaluation.Evaluator`4.Evaluate() [D:\a\1\s\src\LibVLCSharp.Uno\LibVLCSharp.Uno.csproj]
```

mfkl commented:
See https://github.com/videolan/libvlcsharp/commit/00952b64c57e1e2cc65e9a2e4b417e797654a1ff :)

mfkl commented:
I'll have a look this week.

mfkl commented on src/LibVLCSharp.Uno/VideoView.iOS.cs:
```diff
@@ -14,7 +14,7 @@ public partial class VideoView : VideoViewWrapper<Platforms.iOS.VideoView>
         {
             var underlyingVideoView = new Platforms.iOS.VideoView();
-            Border!.Child = underlyingVideoView;
+            Border!.AddSubview( underlyingVideoView);
```
unnecessary extra space

mfkl commented on src/LibVLCSharp/LibVLCSharp.csproj:
```diff
@@ -29,7 +29,7 @@ If you need Xamarin.Forms support, see LibVLCSharp.Forms.
 LibVLC needs to be installed separately, see VideoLAN.LibVLC.* packages.</Description>
     <TargetFrameworks>netstandard2.1;netstandard2.0;netstandard1.1</TargetFrameworks>
-    <TargetFrameworks Condition="!$([MSBuild]::IsOsPlatform('Linux'))">$(TargetFrameworks);MonoAndroid81;Xamarin.iOS10;Xamarin.Mac20;Xamarin.TVOS10</TargetFrameworks>
+    <TargetFrameworks Condition="!$([MSBuild]::IsOsPlatform('Linux'))">$(TargetFrameworks);MonoAndroid9.0;Xamarin.iOS10;Xamarin.Mac20;Xamarin.TVOS10</TargetFrameworks>
```
`MonoAndroid81` won't be enough? Maybe we should provide both `MonoAndroid81` and `MonoAndroid9.0`?

mitchcapper replied:
Sorry, Mono 81 will be enough,  I went and reverted everything else but must have manually downgraded that version I will switch it.

mfkl replied:
Ok

mfkl commented on src/LibVLCSharp.Uno/LibVLCSharp.Uno.csproj:
```diff
@@ -3,7 +3,7 @@
     <Title>LibVLCSharp.Uno</Title>
     <Summary>Uno integration for LibVLCSharp</Summary>
-    <TargetFrameworks>Xamarin.iOS10;MonoAndroid81</TargetFrameworks>
+    <TargetFrameworks>Xamarin.iOS10;MonoAndroid10.0</TargetFrameworks>
```
Uno 3 requires `MonoAndroid10.0` minimum?

mitchcapper replied:
OK well as not an android developer I didn't fully understand what their requirement was when they moved to android 29 as the min SDK version.   Yes MonoAndroid10/API 29 is the minimum for uno, however the "minSdkVersion" for android is Android 5 or  API 21 (per: https://platform.uno/docs/articles/getting-started/requirements.html).   

Revisiting the earlier discussion if the target sdk does not effect the device minimum do we want to upgrade all projects to 10/11 as only effecting what sdk the user compiling has?

Interestingly the minSdkVersion is not possible to set on any android library only on the final project.  so we only set minSdkVersion on 
`samples/Uno/LibVLCSharp.Uno.Sample.Droid`
`samples/Uno/Sample.MediaPlayerElement/Sample.MediaPlayerElement.Droid`

I will change these to 21, but really any user won't actually know the minSdkVersion.

Google actually requires any app updates or new app submissions to target v29 (again no minimum though).   Starting in august that is v30 for new apps and november v30 for existing.   Uno does support MonoAndroid11 right now but you could wait a few months to make the sdk change there.

It is funny without the ability to specify a minSdkVersion on a lib the app will just crash if it is run on too low of a device when that api is used.

mfkl replied:
> Revisiting the earlier discussion if the target sdk does not effect the device minimum do we want to upgrade all projects to 10/11 as only effecting what sdk the user compiling has?

Samples I don't mind. LibVLCSharp.Uno can target MonoAndroid10. But older non-Uno apps may still use MonoAndroid81 so we should keep it on LibVLCSharp.csproj IMHO.

mitchcapper commented:
> See [00952b6](https://github.com/videolan/libvlcsharp/commit/00952b64c57e1e2cc65e9a2e4b417e797654a1ff) :)

Ha!  I saw you added that but as I was updating everything to 3.x I didn't even check what it was referencing.  Clearly yes the same issue.  As I can also repro locally I will make sure to use the latest uno builds and see if it still happens and post to them otherwise.

mitchcapper commented:
One side perk is now it is on 3.8.6 for uno.

mfkl commented:
Updated my system to try to repro the failure from [CI](https://videolan.visualstudio.com/73ec5719-cee8-4da9-937a-9d8a3993d7f5/_apis/build/builds/4411/logs/28), but I can't repro locally, builds fine here.

```
2021-06-16T00:22:10.5924892Z   C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(1961,3): error XA3001: Could not AOT the assembly: LibVLCSharp.Android.AWindow.dll [D:\a\1\s\samples\Uno\LibVLCSharp.Uno.Sample.Droid\LibVLCSharp.Uno.Sample.Droid.csproj]
```

Seems to be a recent tooling issue. Should try this work around https://github.com/xamarin/xamarin-android/issues/5764#issuecomment-856062396

mfkl commented on samples/Uno/Sample.MediaPlayerElement/Sample.MediaPlayerElement.iOS/Sample.MediaPlayerElement.iOS.csproj:
```diff
@@ -114,7 +114,7 @@
   </ItemGroup>
   <ItemGroup>
-    <PackageReference Include="Uno.UI" Version="2.0.527" />
+    <PackageReference Include="Uno.UI" Version="3.7.3" />
```
you meant `3.8.6` here too, I think? :-)

mitchcapper commented:
> Seems to be a recent tooling issue. Should try this work around [xamarin/xamarin-android#5764 (comment)](https://github.com/xamarin/xamarin-android/issues/5764#issuecomment-856062396)

Should that be in aWindow referenced or I would assume the android csproj that is failing.

mfkl commented:
> Should that be in aWindow referenced or I would assume the android csproj that is failing.

I'd have guessed the AWindow bindings project. Though the implications of that workaround could be critical

> A boolean property that controls whether the linker will insert GC.KeepAlive() invocations within binding projects to prevent premature object collection. Defaults to True for Release configuration builds.

Anyway, I tried it in my fork and the build times out https://videolan.visualstudio.com/73ec5719-cee8-4da9-937a-9d8a3993d7f5/_apis/build/builds/4436/logs/28. I cannot reproduce this build failure locally, everything builds fine.

mfkl commented:
Closing for inactivity. Feel free to reopen!

mfkl commented:
```
"D:\a\1\s\src\LibVLCSharp.sln" (Build target) (1) ->
"D:\a\1\s\samples\Uno\Sample.MediaPlayerElement\Sample.MediaPlayerElement.Droid\Sample.MediaPlayerElement.Droid.csproj" (default target) (33) ->
(_LinkAssembliesShrink target) -> 
  C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Xamarin\Android\Xamarin.Android.Legacy.targets(647,5): error XALNK7000: Mono.Linker.MarkException: Error processing method: 'Windows.UI.Xaml.Style LibVLCSharp.Uno.GlobalStaticResources::get___ImplicitStyle_LibVLCSharp_Uno_VideoView()' in assembly: 'LibVLCSharp.Uno.dll' ---> Mono.Cecil.ResolutionException: Failed to resolve Windows.UI.Xaml.DependencyProperty Windows.UI.Xaml.Controls.Control::TemplateProperty [D:\a\1\s\samples\Uno\Sample.MediaPlayerElement\Sample.MediaPlayerElement.Droid\Sample.MediaPlayerElement.Droid.csproj]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Xamarin\Android\Xamarin.Android.Legacy.targets(647,5): error XALNK7000:    at Mono.Linker.Steps.MarkStep.HandleUnresolvedField(FieldReference reference) [D:\a\1\s\samples\Uno\Sample.MediaPlayerElement\Sample.MediaPlayerElement.Droid\Sample.MediaPlayerElement.Droid.csproj]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Xamarin\Android\Xamarin.Android.Legacy.targets(647,5): error XALNK7000:    at Mono.Linker.Steps.MarkStep.MarkInstruction(Instruction instruction) [D:\a\1\s\samples\Uno\Sample.MediaPlayerElement\Sample.MediaPlayerElement.Droid\Sample.MediaPlayerElement.Droid.csproj]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Xamarin\Android\Xamarin.Android.Legacy.targets(647,5): error XALNK7000:    at Mono.Linker.Steps.MarkStep.MarkMethodBody(MethodBody body) [D:\a\1\s\samples\Uno\Sample.MediaPlayerElement\Sample.MediaPlayerElement.Droid\Sample.MediaPlayerElement.Droid.csproj]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Xamarin\Android\Xamarin.Android.Legacy.targets(647,5): error XALNK7000:    at Mono.Linker.Steps.MarkStep.ProcessMethod(MethodDefinition method) [D:\a\1\s\samples\Uno\Sample.MediaPlayerElement\Sample.MediaPlayerElement.Droid\Sample.MediaPlayerElement.Droid.csproj]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Xamarin\Android\Xamarin.Android.Legacy.targets(647,5): error XALNK7000:    at Mono.Linker.Steps.MarkStep.ProcessQueue() [D:\a\1\s\samples\Uno\Sample.MediaPlayerElement\Sample.MediaPlayerElement.Droid\Sample.MediaPlayerElement.Droid.csproj]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Xamarin\Android\Xamarin.Android.Legacy.targets(647,5): error XALNK7000:    --- End of inner exception stack trace --- [D:\a\1\s\samples\Uno\Sample.MediaPlayerElement\Sample.MediaPlayerElement.Droid\Sample.MediaPlayerElement.Droid.csproj]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Xamarin\Android\Xamarin.Android.Legacy.targets(647,5): error XALNK7000:    at Mono.Linker.Steps.MarkStep.ProcessQueue() [D:\a\1\s\samples\Uno\Sample.MediaPlayerElement\Sample.MediaPlayerElement.Droid\Sample.MediaPlayerElement.Droid.csproj]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Xamarin\Android\Xamarin.Android.Legacy.targets(647,5): error XALNK7000:    at Mono.Linker.Steps.MarkStep.ProcessPrimaryQueue() [D:\a\1\s\samples\Uno\Sample.MediaPlayerElement\Sample.MediaPlayerElement.Droid\Sample.MediaPlayerElement.Droid.csproj]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Xamarin\Android\Xamarin.Android.Legacy.targets(647,5): error XALNK7000:    at Mono.Linker.Steps.MarkStep.Process() [D:\a\1\s\samples\Uno\Sample.MediaPlayerElement\Sample.MediaPlayerElement.Droid\Sample.MediaPlayerElement.Droid.csproj]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Xamarin\Android\Xamarin.Android.Legacy.targets(647,5): error XALNK7000:    at Mono.Linker.Steps.MarkStep.Process(LinkContext context) [D:\a\1\s\samples\Uno\Sample.MediaPlayerElement\Sample.MediaPlayerElement.Droid\Sample.MediaPlayerElement.Droid.csproj]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Xamarin\Android\Xamarin.Android.Legacy.targets(647,5): error XALNK7000:    at MonoDroid.Tuner.MonoDroidMarkStep.Process(LinkContext context) [D:\a\1\s\samples\Uno\Sample.MediaPlayerElement\Sample.MediaPlayerElement.Droid\Sample.MediaPlayerElement.Droid.csproj]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Xamarin\Android\Xamarin.Android.Legacy.targets(647,5): error XALNK7000:    at Mono.Linker.Pipeline.ProcessStep(LinkContext context, IStep step) [D:\a\1\s\samples\Uno\Sample.MediaPlayerElement\Sample.MediaPlayerElement.Droid\Sample.MediaPlayerElement.Droid.csproj]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Xamarin\Android\Xamarin.Android.Legacy.targets(647,5): error XALNK7000:    at Mono.Linker.Pipeline.Process(LinkContext context) [D:\a\1\s\samples\Uno\Sample.MediaPlayerElement\Sample.MediaPlayerElement.Droid\Sample.MediaPlayerElement.Droid.csproj]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Xamarin\Android\Xamarin.Android.Legacy.targets(647,5): error XALNK7000:    at MonoDroid.Tuner.Linker.Process(LinkerOptions options, ILogger logger, LinkContext& context) [D:\a\1\s\samples\Uno\Sample.MediaPlayerElement\Sample.MediaPlayerElement.Droid\Sample.MediaPlayerElement.Droid.csproj]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Xamarin\Android\Xamarin.Android.Legacy.targets(647,5): error XALNK7000:    at Xamarin.Android.Tasks.LinkAssemblies.Execute(DirectoryAssemblyResolver res) [D:\a\1\s\samples\Uno\Sample.MediaPlayerElement\Sample.MediaPlayerElement.Droid\Sample.MediaPlayerElement.Droid.csproj]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Xamarin\Android\Xamarin.Android.Legacy.targets(647,5): error XALNK7000:    at Xamarin.Android.Tasks.LinkAssemblies.RunTask() [D:\a\1\s\samples\Uno\Sample.MediaPlayerElement\Sample.MediaPlayerElement.Droid\Sample.MediaPlayerElement.Droid.csproj]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Xamarin\Android\Xamarin.Android.Legacy.targets(647,5): error XALNK7000:    at Microsoft.Android.Build.Tasks.AndroidTask.Execute() in /Users/builder/azdo/_work/1/s/xamarin-android/external/xamarin-android-tools/src/Microsoft.Android.Build.BaseTasks/AndroidTask.cs:line 17 [D:\a\1\s\samples\Uno\Sample.MediaPlayerElement\Sample.MediaPlayerElement.Droid\Sample.MediaPlayerElement.Droid.csproj]


"D:\a\1\s\src\LibVLCSharp.sln" (Build target) (1) ->
"D:\a\1\s\samples\Uno\LibVLCSharp.Uno.Sample.Droid\LibVLCSharp.Uno.Sample.Droid.csproj" (default target) (34) ->
  C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Xamarin\Android\Xamarin.Android.Legacy.targets(647,5): error XALNK7000: Mono.Linker.MarkException: Error processing method: 'Windows.UI.Xaml.Style LibVLCSharp.Uno.GlobalStaticResources::get___ImplicitStyle_LibVLCSharp_Uno_VideoView()' in assembly: 'LibVLCSharp.Uno.dll' ---> Mono.Cecil.ResolutionException: Failed to resolve Windows.UI.Xaml.DependencyProperty Windows.UI.Xaml.Controls.Control::TemplateProperty [D:\a\1\s\samples\Uno\LibVLCSharp.Uno.Sample.Droid\LibVLCSharp.Uno.Sample.Droid.csproj]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Xamarin\Android\Xamarin.Android.Legacy.targets(647,5): error XALNK7000:    at Mono.Linker.Steps.MarkStep.HandleUnresolvedField(FieldReference reference) [D:\a\1\s\samples\Uno\LibVLCSharp.Uno.Sample.Droid\LibVLCSharp.Uno.Sample.Droid.csproj]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Xamarin\Android\Xamarin.Android.Legacy.targets(647,5): error XALNK7000:    at Mono.Linker.Steps.MarkStep.MarkInstruction(Instruction instruction) [D:\a\1\s\samples\Uno\LibVLCSharp.Uno.Sample.Droid\LibVLCSharp.Uno.Sample.Droid.csproj]
```

mfkl commented:
Now there is a 4.x Uno stable version..

mfkl commented:
Closing for inactivity.
